sdks/python/apache_beam/io/gcp/bigquery.py

```diff
@@ -2525,6 +2526,12 @@ def _get_pipeline_details(unused_elm):
                 **self._kwargs))
         | _PassThroughThenCleanupTempDatasets(project_to_cleanup_pcoll))
 
+  def get_pcoll_from_schema(table_schema):
+    pcoll_val = apache_beam.io.gcp.bigquery_schema_tools.\
+        produce_pcoll_with_schema(table_schema)
+    return beam.Map(lambda values: pcoll_val(**values)).with_output_types(
```

Comment:
@TheNeuralBit , tried this here but looks like the pickling still had an error. Something to note is that this is able to run fine on Colab notebook via DirectRunner (https://colab.sandbox.google.com/drive/1_BkzBwCLB7OjR1jGGzCkxXFHreUZgCZl#revisionId=0BzxzuriHpr5fcXNOWHcvcGlqYTBWWUZxRWdQaFIrVVFYV0Y4PQ&scrollTo=YITcZHkoZR59) 

Is there a way we can further see which coder is being used?

Reply:
Sure, the reason we shouldn't be pickling instances of this type is because we want them to be encoded with SchemaCoder (rather than the default PickleCoder) when they're used as the element_type of a PCollection

Reply:
In our sync today let's take a look at what coder is being used for the result pcollection, it looks like it's not resolving to SchemaCoder for some reason

Reply:
When we met we weren't able to repro the issue in the test locally because of other auth issues. We did run through some logic like the test in a colab notebook, and it didn't have any problem. When we observed the pcolleciton in that case it had the element_type set properly, and we verified that that element_type was registered to use RowCoder.

I think we need to either get the test running locally and check those same data points, or do that via print() debugging on jenkins (as a last resort).

Reply:
Looks like this was using ``rowcoder``` and the element type was as expected. The error itself is being thrown from [here](https://github.com/svetakvsundhar/beam/blob/bqio/sdks/python/apache_beam/coders/coders.py#L864), could this be because of the protocol being used? I tried using ```DEFAULT_PROTOCOL```, but that didn't seem to work..could you point me to the specific rowcoder implementation?

Reply:
@TheNeuralBit just to follow up here, it looks like ```rowcoder``` was indeed the coder being used. I verified that the asserts passed here https://github.com/svetakvsundhar/beam/blob/bqio/sdks/python/apache_beam/coders/row_coder_test.py#L263, and they both proved to be using rowcoder.

Are there any other intermediate methods I can run on our ```usertype``` to verify during the rowcoder ptransform? One thing I noted was that ```usertype``` doesn't have a ```.schema``` property, and that I wasn't able to verify if the argument of this ptransform is a schema (```apache_beam.portability.api.schema_pb2.Schema```). Is there an easy way to verify that the input is indeed of that type?

Reply:
Ah ok I think I've found the RC: https://cloud.google.com/dataflow/docs/resources/faq#how_do_i_handle_nameerrors

Looks like the pickling is happen in my global namespace, and not carrying over to the Dataflow worker. A fix such as passing in ```--save_main_session``` should do the trick. See https://issues.apache.org/jira/browse/BEAM-6158 for a similar error that is tangentially related.